virtcontainers/kata_agent.go

```diff
@@ -962,3 +963,22 @@ func (k *kataAgent) sendReq(request interface{}) (interface{}, error) {
 		return nil, fmt.Errorf("Unknown gRPC type %T", req)
 	}
 }
+
+// ParseConfigJSON unmarshals the config.json file.
+// TODO: this is duplicate with virtcontainers/pkg/oci/utilsgo:ParseConfigJSON
+// but we can't use that function because there will be cycle dependency problem
+// better find a way to eliminate duplicate codes @weizhang555
+func ParseConfigJSON(bundlePath string) (*specs.Spec, error) {
+	configPath := filepath.Join(bundlePath, "config.json")
+	configByte, err := ioutil.ReadFile(configPath)
```

Comment:
With this, we are pushing the cli dependence on `config.json` file to virtcontainers. If we are certain that the content of OCI annotations is exactly the same as `config.json`, and assuming the purpose of the PR is to reduce persistent data, can we still let the cli read `config.json` and pass its contents to virtcontainers, and skip persisting the OCI spec in virtcontainers?

Reply:
Then it will need to filter out "configJSON" content overtime we try to persist the data, it could be ugly codes and everywhere.

Are you worrying about the operation performance? I think it will influence a little on performance because it will need to open files more times, but the influence won't be much in my opinion. Though I don't have a test data. 😄

Reply:
@WeiZhang555 I'm thinking about the following things:
1. We do not change the image bundle's config.json in hyperd for a container one. I think we can deal with it if required.
2. This requires `config.json` to be exactly as what's in the annotations. Are we certain about it?
3. If we ever change the OCI spec in virtcontainers, do we update the `config.json` file?

Reply:
I didn't fully get your concerns. But in theory, I think it's better virtcontainers do never change `config.json` file because it's an input from upper level component or user, and we can add custom or additional modifications into virtcontainer's persist data but not original bundle config.json file.
Is this what you want to discuss?

Reply:
Yes, that's it. We do change the OCI spec in certain cases (e.g., #138) and we do need to persist the new OCI spec in such case to allow future fetchSandbox() caller to get the same spec view of the container. I agree with you that we should **NOT** change the contents of `config.json`. So it means we need to find a way to persist optional modification to the spec and merge with `config.json` when reading it. If you do not want to handle it in this PR, we need to file an issue to make sure it gets implemented later.

Reply:
@bergwolf I didn't find why we need to change OCI spec in #138, I think we can support `-v /dev/block-dev:xxx` with standard OCI spec. 
Can you point me the dedicated discussion about this? Or some more description?

Reply:
@WeiZhang555 e.g. https://github.com/kata-containers/runtime/pull/138/files#diff-7ba6ce1174401126aaba6a1cff25434dR559

Reply:
@bergwolf This is a good point!  I didn't realize we have such logic.

But even in previous code, after we modified the spec, the new spec data isn't persisted or saved back to annotation, so the good point is this PR didn't influence any of the function.

Reply:
@WeiZhang555 I think we need to fix it before releasing. It is quite easy to fix before this PR. If we merge the PR, it won't be straightforward to fix any more.

Reply:
Do you mean we need to persist the modified spec?

I mean the modified spec is calculated everytime you send a request to kata-agent, so in current logic, there's nothing need to be fixed.

Reply:
No, it's only calculated when we create a new container (look for callers of `replaceOCIMountsForStorages`). Later API calls only read the OCI spec from persistent data and that is not the same one we have sent to the agent.

Reply:
Here I'm still confused on what to fix.

Here when we modified the OCI spec, we want to send a temporary modified spec to kata-agent, but we don't want to persist the modified ones. So bundle/config.json will not be modified in any cases, and reading from bundle/config.json should be right logic as this PR is doing. Right?

So what should I fix here?

Reply:
>  but we don't want to persist the modified ones

That's what I wanted to fix. If we decide not to persist the modified spec, that's a design decision and we need to be really careful handling it because we are not recording what's sent to the agent and other virtcontainers API calls are getting different OCI spec than the one sent to the agent.

Reply:
Fair enough. `createContainer()` in kata_agent.go seems to be the only user of the OCI spec in virtcontainers. I might just be over worried.

Another thing is container hooks. If we read the spec so late, how do we handle pre-start hooks?

Reply:
@bergwolf for container hooks, it will be stored inside SandboxConfig from OCI spec, so when the sandbox runs the hooks, it doesn't need to read OCI spec again.

https://github.com/kata-containers/runtime/blob/master/virtcontainers/pkg/oci/utils.go#L494

Reply:
oops, sorry I missed it. We are reading the OCI spec file twice, one in cli.Create and the other in virtcontainers.createContainer.